src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,445 @@
+exprt constant_float(const double f, const ieee_float_spect &float_spec)
+{
+  ieee_floatt fl(float_spec);
+  if(float_spec==ieee_float_spect::single_precision())
+  {
+    fl.from_float(f);
+  }
+  else
+  {
+    assert(float_spec==ieee_float_spect::double_precision());
+    fl.from_double(f);
+  }
+  return fl.to_expr();
+}
+
+/// Round a float expression to an integer closer to 0
+/// \param f: expression representing a float
+/// \return expression representing a 32 bit integer
+exprt round_expr_to_zero(const exprt &f)
+{
+  exprt rounding=from_integer(ieee_floatt::ROUND_TO_ZERO, unsignedbv_typet(32));
+  return floatbv_typecast_exprt(f, rounding, signedbv_typet(32));
+}
+
+/// Multiplication of expressions representing floating points.
+/// Note that the rounding mode is set to ROUND_TO_EVEN.
+/// \param f: a floating point expression
+/// \param g: a floating point expression
+/// \param rounding: rounding mode
+/// \return An expression representing floating point multiplication.
+exprt floatbv_mult(const exprt &f, const exprt &g)
+{
+  ieee_floatt::rounding_modet rounding=ieee_floatt::ROUND_TO_EVEN;
+  assert(f.type()==g.type());
+  exprt mult(ID_floatbv_mult, f.type());
+  mult.copy_to_operands(f);
+  mult.copy_to_operands(g);
+  mult.copy_to_operands(from_integer(rounding, unsignedbv_typet(32)));
+  return mult;
+}
+
+/// Convert integers to floating point to be used in operations with
+/// other values that are in floating point representation.
+/// \param i: an expression representing an integer
+/// \param spec: specification for floating point numbers
+/// \return An expression representing representing the value of the input
+///   integer as a float.
+exprt floatbv_of_int_expr(const exprt &i, const ieee_float_spect &spec)
+{
+  exprt rounding=from_integer(ieee_floatt::ROUND_TO_ZERO, unsignedbv_typet(32));
+  return floatbv_typecast_exprt(i, rounding, spec.to_type());
+}
+
+/// Estimate the decimal exponent that should be used to represent a given
+/// floating point value in decimal.
+/// We are looking for d such that n * 10^d = m * 10^e, so:
+/// d = log_10(m) + log_10(2) * e - log_10(n)
+/// m -- the magnitude -- should be between 1 and 2 so log_10(m)
+/// in [0,log_10(2)].
+/// n -- the magnitude in base 10 -- should be between 1 and 10 so
+/// log_10(n) in [0, 1].
+/// So the estimate is given by:
+/// d ~=~  floor( log_10(2) * e)
+/// \param f: a floating point expression
+/// \param spec: specification for floating point
+exprt estimate_decimal_exponent(const exprt &f, const ieee_float_spect &spec)
+{
+  exprt log_10_of_2=constant_float(0.30102999566398119521373889472449302, spec);
+  exprt bin_exp=get_exponent(f, spec);
+  exprt float_bin_exp=floatbv_of_int_expr(bin_exp, spec);
+  exprt dec_exp=floatbv_mult(float_bin_exp, log_10_of_2);
+  binary_relation_exprt negative_exp(dec_exp, ID_lt, constant_float(0.0, spec));
+  // Rounding to zero is not correct for negative number, so we remove 1.
```

Comment:
If I remember well, when I tried with the correct one, I would get an exception from CBMC because it is not implemented.